Adding ability to get status from Treeherder by --query-source flag.

vaibhavmagarwal opened the pull request.

vaibhavmagarwal commented:
r? @armenzg

vaibhavmagarwal commented:
I have tested the code for this case:
    python trigger.py --rev b4fd9967aa85 -b "Ubuntu HW 12.04 x64 fx-team talos g1 " --times 3 --status-provider "treeherder"

and it works out fine. Also, a lot of time is saved by not downloading builds-4hr.js via this approach as compared to:
    python trigger.py --rev b4fd9967aa85 -b "Ubuntu HW 12.04 x64 fx-team talos g1 " --times 3

vaibhavmagarwal commented:
I tried triggering with --status-provider "treeherder" flag and it has been successfull. Examples:
https://treeherder.mozilla.org/#/jobs?repo=mozilla-inbound&fromchange=ba3045335448&tochange=ba3045335448&filter-searchStr=Ubuntu+HW+12.04+mozilla-inbound+talos+g1

https://treeherder.mozilla.org/#/jobs?repo=mozilla-inbound&fromchange=ba3045335448&tochange=ba3045335448&filter-searchStr=Ubuntu%20VM%2012.04%20x64%20mozilla-inbound%20opt%20test%20mochitest-e10s-browser-chrome-1

armenzg commented on mozci/mozci.py:
```diff
@@ -64,7 +64,10 @@ def _status_summary(jobs):
+        if status_provider == "treeherder":
+            status = treeherder.query_job_status(job)
+        else:
+            status = buildapi.BuildapiJobStatus(job).get_status()
```
We should not have to use status_provider optional flag.
We should have a global StatusProvider base class which we use in this function.
The function becomes oblivious to what provider is being used underneath.
We should initialize this status provider somewhere else.
Probably in the script.
It will remove the need to status_provider to lots of functions and if/else conditions.

armenzg commented on mozci/mozci.py:
```diff
@@ -64,7 +64,10 @@ def _status_summary(jobs):
+            status = treeherder.query_job_status(job)
+        else:
+            status = buildapi.BuildapiJobStatus(job).get_status()
         if status == buildapi.PENDING:
```
In here, we're comparing against **buildapi**.PENDING.
If TH and BuildAPI share common codes, we can probably have a jobs.py or status.py module (or even within mozci.py) that contains these constants.

What do you think?

armenzg replied:
It could even be part of the main StatusProvider base class.

Note that you will likely have to re-evaluate if we implemented BuildapiJobStatus() properly. I think so but I will let you think about it.

armenzg commented on mozci/mozci.py:
```diff
@@ -384,9 +388,14 @@ def trigger_range(buildername, revisions, times=1, dry_run=False, files=None):
+        else:
+            jobs = buildapi.query_jobs_schedule(repo_name, rev)
+            matching_jobs = _matching_jobs(buildername, jobs)
+        successful_jobs, pending_jobs, running_jobs = _status_summary(matching_jobs,
```
Similar comment as above.

armenzg commented on mozci/mozci.py:
```diff
@@ -405,6 +414,10 @@ def trigger_range(buildername, revisions, times=1, dry_run=False, files=None):
             # instead of creating a new arbitrary job
+            if status_provider == "treeherder":
+                jobs = buildapi.query_jobs_schedule(repo_name, rev)
+                matching_jobs = _matching_jobs(buildername, jobs)
```
The comment above this new code should be below.

Why do you check for status again in here?
It is unlikely that the status will change from the previous request just a second ago.

vaibhavmagarwal replied:
I need to check the status from buildapi here if we have been getting from "treeherder" because we use "requests" key and "request_id" key for triggering the job below and it is not present in the job dict from treeherder. So I had to get it from buildapi itself.

adusca replied:
It is possible to get the request_id from a TH job_id: https://github.com/adusca/pulse_actions/blob/master/pulse_actions/utils/treeherder.py#L6

vaibhavmagarwal replied:
Thanks for the info @adusca , I'll see how this will fit in since in the new model I am planning to make the function oblivious to what provider is being used underneath. But hopefully this will come in handy!

armenzg commented on mozci/mozci.py:
```diff
@@ -465,9 +478,14 @@ def backfill_revlist(buildername, revisions):
+            matching_jobs = treeherder._matching_jobs(buildername, jobs)
+        else:
+            jobs = buildapi.query_jobs_schedule(repo_name, rev)
+            matching_jobs = _matching_jobs(buildername, jobs)
```
Same idea as above.

armenzg commented:
Excellent job! I'm glad we're finally gaining this feature.

Could you please ask someone else with your th-client knowledge to review your usage of th client?

vaibhavmagarwal commented on mozci/scripts/trigger.py:
```diff
@@ -63,6 +63,12 @@ def parse_args(argv=None):
+                        default="buildapi",
+                        help="Query info from buildapi or treeherder.")
+
     parser.add_argument("--file",
```
I have added --query-source instead of --status-provider because we are querying treeherder api for other things too (like jobs schedule, status etc). So --query-source flag made more sense to me, let me know if you think otherwise.

armenzg replied:
WFM.

vaibhavmagarwal commented on mozci/sources/buildapi.py:
```diff
@@ -195,71 +194,6 @@ def valid_revision(repo_name, revision):
 #
 # Functions to query
 #
-class BuildapiJobStatus(object):
```
Had to move BuildapiJobStatus methods to query_jobs.py since we needed the job status from treeherder and buildapi in a same class, abstracted from the user.

vaibhavmagarwal commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,176 @@
+LOG = logging.getLogger('mozci')
+PENDING, RUNNING, COALESCED, UNKNOWN = range(-4, 0)
+SUCCESS, WARNING, FAILURE, SKIPPED, EXCEPTION, RETRY, CANCELLED = range(7)
+HOST_ROOT = 'https://secure.pub.build.mozilla.org/buildapi/self-serve'
```
I can have SUCCESS, RUNNING, COALESCED etc. as class attributes of QueryJob class, but I am not sure, so I left it as it is. What are your thoughts?

armenzg replied:
Let's leave it here for now.

vaibhavmagarwal commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,176 @@
+
+
+class QueryJob(object):
+
```
I named the class QueryJob instead of StatusProvider because we are having other methods like get_all_jobs (for a given revision), get_matching_jobs apart from get_job_status. So, I have given a generic name. If you can come up with better class names, let me know.

armenzg replied:
QueryApi ?
I'm fine either way.

vaibhavmagarwal commented on test/test_buildapi.py:
```diff
@@ -78,46 +78,6 @@ def mock_response_json():
 
 
-class TestQueryJobsSchedule(unittest.TestCase):
-
```
I have removed this test for time being since the functions were moved out of buildapi. Once I know if my approach is correct, I will add these test functions to the appropriate file.

armenzg replied:
Let's not forget of adding these tests back now that you know that the direction is looking right.
Otherwise, we are likely to regress something.

armenzg replied:
It's hard to easily see since some tests have been moved to test_query_jobs.py
I believe you have managed to not miss any functions in the porting.

vaibhavmagarwal replied:
Well, I have shifted all of them, as you can see a 0.0% change in the coverage by coveralls

vaibhavmagarwal commented:
r? @armenzg
I have made some decisions with regards to class design and would like your feedback on it. Also, for time being I have removed some test functions due to shifting of methods, and will add them back (to an appropriate file) after getting your feedback. Thanks

armenzg commented on mozci/mozci.py:
```diff
@@ -55,7 +42,7 @@ def _unique_build_request(buildername, revision):
 
 
-def _status_summary(jobs):
+def _status_summary(jobs, query_source="buildapi"):
```
Instead of definying query_source="buildapi" on every function.
You can instanciate one global object which we can access at all times.

You can have two classes extending QueryJob.

You can have this code:
```
if query_source="buildapi":
     source_class = BuildApi
else:
     source_class = TreeherderApi

QUERY_SOURCE = source_class()
```

If you need to not use the QUERY_SOURCE instance, you can create other instances where needed.

armenzg commented:
Hi @vaibhavmagarwal 
My main request is to split into two classes that way we can take a more optimized usage of OOP.
If you want branch from here to keep this changes at hand, however, I'm pretty certain that I prefer having the two classes extending the base class.

Thanks for your hard work and let me know if I need to clarifiy further.

I will review the patch closely once I see this change happening.

vaibhavmagarwal commented:
r? @armenzg 
Armen, this is a big patch, and although I have gone through the files several times, there may be something that I might have missed. Apologies in advance!
Unfortunately, since a lot has changed in this patch, the unit tests no longer can stay in the same file, they need to be written again and in a different file. I will see how it can be done, once I get a review from you and know the correct approach.

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -223,6 +221,18 @@ def _find_files(job_schedule_info):
+    all_jobs = query_api.get_all_jobs(repo_name, revision)
+    return [job["requests"][0]["request_id"] for job in all_jobs
+            if query_api.get_job_status(job) == status]
+
```
Needed to move this function from buildapi.py since it causes a circular dependency between buildapi.py and query_jobs.py

armenzg replied:
Couldn't this be part of the BuildApi class?

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -367,8 +378,10 @@ def trigger_job(revision, buildername, times=1, files=None, dry_run=False,
+                  dry_run=False, files=None):
     """Schedule the job named "buildername" ("times" times) in every revision on 'revisions'."""
+    set_query_source(query_source)
     repo_name = query_repo_name_from_buildername(buildername)
```
After thinking on this a bit, I think that the call to set_query_source() function should be called from trigger_range instead of user script to minimize the number of things to be kept care of while writing a user facing script. Let me know if I should change this.

armenzg replied:
This would be OK if alltalos did not call directly trigger_job() instead of trigger_range().

Perhaps, we can add a wrapper to those two functions. I'm not sure if it would be the right way.
https://wiki.python.org/moin/FunctionWrappers

We can also leave it up to the script to decide if they're to use TH status or not.

armenzg replied:
As well as backfill.

vaibhavmagarwal commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,170 @@
+    """ Base class for common query methods """
+
+    def __init__(self, query_source):
+        self.query_source = query_source
```
Right now this class is really not needed, but I have kept it as a base class in case we need more generic function common to both BuildApi and TreeherderApi in the future.

armenzg replied:
We need to specify in the base class the set of functions which are common to sub-classing classes. We add "pass" in their body.

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,170 @@
+        artifact_content = requests.get(artifact_url, params=query_params).json()
+        return artifact_content[0]["blob"]["request_id"]
+
+    def get_matching_jobs(self, buildername, all_jobs):
```
Could you please add similar debug info as the buildapi method one?

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,170 @@
+        artifact_url = self.TH_ARTIFACT_URL.format(repo_name)
+        query_params = {'job_id': job_id,
+                        'name': 'buildapi'}
+        artifact_content = requests.get(artifact_url, params=query_params).json()
```
Can you please add a debug message to see what url we are going to hit?

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -223,6 +221,17 @@ def _find_files(job_schedule_info):
+    all_jobs = QUERY_SOURCE.get_all_jobs(repo_name, revision)
+    return [QUERY_SOURCE.get_buildapi_request_id(repo_name, job) for job in all_jobs
+            if QUERY_SOURCE.get_job_status(job) == status]
+
```
find_all_by_status function right now is being used for Coalesced jobs, but I think we can make it a generic useful function. @adusca , is this conversion ok? or do I need to change something?

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -369,6 +382,9 @@ def trigger_job(revision, buildername, times=1, files=None, dry_run=False,
     """Schedule the job named "buildername" ("times" times) in every revision on 'revisions'."""
+    if QUERY_SOURCE is None:
+        set_query_source()
+
```
The above code is to ensure, that one does not need to call set_query_source() is he/she does not need TH status in their script.

armenzg replied:
You can do this as a decorator which will keep these functions a bit cleaner.
Also remember that you need to call "global QUERY_SOURCE" first.

This is an example I wrote:
#! /usr/bin/env python
from bar import C1, C2
SOURCE = "buildapi"
OBJ = None

def set_source(source):
    global SOURCE
    SOURCE = source

class query_source(object):
    def __init__(self, f):
        self.f = f

    def __call__(self):
        global SOURCE
        global OBJ
        print "Setting OBJ"
        if SOURCE == "buildapi":
            OBJ = C1()
        else:
            OBJ = C2()

        self.f()

@query_source
def my_func():
    global OBJ
    OBJ.make_noise()

my_func()
set_source("treeherder")
my_func()
MacAir temp $ python foo.py
Setting OBJ
moo
Setting OBJ
quack

vaibhavmagarwal commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,195 @@
+
+    def __init__(self, query_source):
+        self.query_source = query_source
+
```
Added "abstractmethods" below like told, but I am not sure if they serve any purpose.

armenzg replied:
When we define abstract methods, we make a promise to consumers of the code.
As you indicate, there is no functional value, however, if you try to extend this class and do *not* implement all of the methods marked as abstract, the derived class would be *incomplete* or not fullfillyng all promises.

Do you know what I mean?

BTW, I like that you found this decorators. I've learned something new today!

vaibhavmagarwal replied:
Heh, I know what an abstract class is :)
I just wanted to make sure that having it made sense here.

vaibhavmagarwal commented:
r?

armenzg commented:
@vaibhavmagarwal this is going in the right direction. Let's add the decorators as I mentioned.
As you been able to test a handful of scripts with both modes? (With --dry-run that is)
I suggest trying on m-i with recent pushes (within the last 4 hours) and older pushes on other trees.

@adusca could you please review this code and see if it is looking good?
It is a big refactor and I want a second opinion.

Also, I'm not yet done. I want to review the new module closely to make sure that I'm not missing anything.

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -15,25 +15,34 @@
+        if QUERY_SOURCE is None:
+            set_query_source()
+        return func(*args, **kwargs)
 
```
While the class method of decorating was good, I think the above is a more often used way of decorating a function: https://docs.python.org/2/library/functools.html#functools.wraps and it also is smaller and cleaner.

armenzg replied:
Yes, this looks prettier!

When you call "global QUERY_SOURCE" in the wrapper, does it become available withing the function wrapped?
That is neat if so!

vaibhavmagarwal commented:
r?
Tested the scripts for both last 4 hours and older pushes on other branches.
[1] https://pastebin.mozilla.org/8837160
[2] https://pastebin.mozilla.org/8837161
[3] https://pastebin.mozilla.org/8837162

adusca commented:
Would it be too messy to move the class BuildApi to buildapi.py? I believe this could make the code easier to navigate.

armenzg commented:
I have never realized how making the right decision for a refactoring can get!!

I see clearly both approaches as valid:
* Everything that is related to buildapi in one module
* Related classes to be on the same module

I don't have enough experience helping in such decissions. I will ask wlach.

I will keep on looking today at this code to see what is the right way.

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,214 @@
+        status_data = query_job_data(req["complete_at"], req["request_id"])
+        return status_data["properties"]["revision"][0:12] != req["revision"][0:12]
+
+    def find_all_jobs_by_status(self, repo_name, revision, status):
```
Is the name of the function correct? I think it should be find_all_by_status()

vaibhavmagarwal replied:
Ahh, nice catch!

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,214 @@
+        raise Exception("Unexpected status")
+
+    def _is_coalesced(self, job):
+        """Helper method to determine if a job with status 'SUCCESS' is coalesced."""
```
Could you please reference this bug in here?
https://bugzilla.mozilla.org/show_bug.cgi?id=1175611

Could you mention that this raises a BuildjsonException?
https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/sources/buildjson.py#L77

Could you also fix the query_job_data() doc to mention BuildjsonException instead of Exception?

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,214 @@
+                return CANCELLED
+
+        LOG.debug(job)
+        raise Exception("Unexpected status")
```
Could you please add a MozciException in mozci.py and use that?

armenzg commented on mozci/sources/query_jobs.py:
```diff
@@ -0,0 +1,214 @@
+                                                  result_set_id=revision_id)
+        return all_jobs
+
+    def get_buildapi_request_id(self, repo_name, job):
```
Should this method live in the treeherder-client?
I don't know but throwing that out there.

vaibhavmagarwal replied:
Good idea! I have sent a PR: https://github.com/mozilla/treeherder/pull/657 for it.

armenzg commented:
NOTE: I won't look at more code reviews until Friday. I can answer questions though and quick code checks.

wlach says "it depends" wrt to my refactoring question (https://github.com/armenzg/mozilla_ci_tools/pull/257#issuecomment-112880028).
He recommended to split it if it gets too big.

I think that my last 3 requests are:
* We need the tests back
* Move query_jobs.py from mozci/sources to mozci/
* Move core functionality back to the buildapi module which the Buildapi class will call

I think there is value to move the core code in BuildApi class into the buildapi.py module.
This allow us to have clear decoupled code. Think on how you let the TH client deal with the system rather than having code in this class to do the job of the client.
The reason behind this is that we almost want buildapi.py to be the closest to be a client in the same way that there is a treeherder client.
The reason we don't have a treeherder module is because we have a client.
Does this make sense to you?

I'm thinking of the internal functionality of get_all_jobs() and maybe find_all_jobs_by_status().

Even though, I originally wanted to keep every interaction with 1 system being in the same module (the sources module). I think it makes sense to split it into functionality.

If we want to query_jobs, we go to the query_jobs module.
* in one case, we use the th client to actually do stuff; in the other, we make calls to the buildapi module
If we want to trigger something, we go to the trigger module (non-existant).
* one day, we will have the Trigger() base class which gets extended; we will call code in buildapi.py
If we want to deal with pushes, we go to the pushes module (non-existant)
* one day, we might have a Push() base class which gets extended for either Hg or Git; in the case of Hg we internally use the pushlog module

vaibhavmagarwal commented:
r? @armenzg 
Armen, I have made the changes you have suggested. Also, please note that I have re-added all the tests which were there previously. I have not yet added tests for TreeherderApi class as they will take sometime to make, and since we are using an optional flag (--query-source) for it, it should not break any existing functionality. I will add those in the next PR. Thanks!

vaibhavmagarwal commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+            return JOBS_CACHE[(repo_name, revision)]
+
+        JOBS_CACHE[(repo_name, revision)] = \
+            buildapi.query_for_jobs(repo_name, revision)
```
The buildapi client functionality has been moved to buildapi.py, and this method calls it.

armenzg replied:
Thanks for fixing it!

You can probably write this code like this (if I got it right):
```
+ if (repo_name, revision) not in JOBS_CACHE:
+        JOBS_CACHE[(repo_name, revision)] = \
+            buildapi.query_for_jobs(repo_name, revision)
+
+ return JOBS_CACHE[(repo_name, revision)]
```

armenzg commented on mozci/mozci.py:
```diff
@@ -16,24 +16,34 @@
-    for j in all_jobs:
-        if j["buildername"] == buildername:
-            matching_jobs.append(j)
+def set_query_source(query_source="buildapi"):
```
@vaibhavmagarwal what do you think if QUERY_SOURCE, set_query_source and query_source_wrapper lived inside of query_jobs.py?

We could only import in here the wrapper ("query_source_wrapper"). That way all logical parts are inside the same module.

vaibhavmagarwal replied:
I think we have to leave this in mozci.py itself, since this module is a generic one. Also, QUERY_SOURCE is a global variable, so we will need to import that too from query_jobs.py and it is used in just this module. So functionality wise I think it makes sense to have these functions here itself.

chmanchester replied:
I agree with @armen on this point. The query_source_wrapper is spilling the question of whether we already know where we're getting data out into the methods of mozci.py when it could be contained in query_jobs.py.

I think a better design would be to have set_query_source in query_jobs.py, and call it only when we need to pass it user configuration (I see this happening in the scripts). Calling set_query_source would set up whatever global state the query_jobs module needs know that when mozci.py calls it it should get data from one or the other source.

armenzg commented:
I will finish reviewing this after lunch. You can wait until I'm done before addressing my first comment.

armenzg commented:
@vaibhavmagarwal my apologies if this is being such a long process.
I've noticed that is a bit hard to see the final product in my head before we go through few iterations.
You've done a great job here and it's cool to see that *I* believe will be more easy to be maintained in the long term.

Remaining comments coming.

armenzg commented on mozci/mozci.py:
```diff
@@ -64,14 +75,14 @@ def _status_summary(jobs):
-        status = buildapi.BuildapiJobStatus(job).get_status()
-        if status == buildapi.PENDING:
+        status = QUERY_SOURCE.get_job_status(job)
+        if status == query_jobs.PENDING:
```
In these two lines we read about QUERY_SOURCE and query_jobs and I wonder what would be easier to read and to make it easier to understand.

When we do "query_jobs.ATTRIBUTE" it looks like we're dealing with an object yet we're dealing with a module.

I prefer explicit imports so this should make it clearer:
+from mozci import query_jobs, PENDING, RUNNING, SUCCESS, COALESCED

armenzg commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+from sources.buildjson import query_job_data, BuildjsonException
+
+
+LOG = logging.getLogger('mozci')
```
Do we need to call setup_logging? I don't think so but I'm not 100% sure.

https://github.com/armenzg/mozilla_ci_tools/commit/48da77c27ab5998b377dd620ef38844f93a56939

adusca replied:
We just need to call setup_logging for command-line scripts, it deals with modules logs automatically.

chmanchester replied:
Just curious -- why aren't you all using mozlog here?

armenzg commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+        results = self.treeherder_client.get_resultsets(repo_name, revision=revision)
+        all_jobs = []
+        if results:
+            revision_id = results[0]["id"]
```
What is the revision_id? Is it an internal ID that represents a revision in TH?
Could you please add a note if so.

I wonder if a one single call method in the TH client would make sense or not. FYI I'm not asking for it, you know the code better and if it would make sense to file that or not.

vaibhavmagarwal replied:
Treeherder-client is a thin wrapper over treeherder api, so it will not make sense to get revision id directly from it, since this is a specific use case, and different projects will require different information from get_resultsets. I will add a comment though for this.

Right now, we are using two calls to treeherder, it will be good if we had a single call to treeherder api, and I have filed a bug for that - https://bugzilla.mozilla.org/show_bug.cgi?id=1165401

armenzg commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+        if job["state"] == "completed":
+            if job["result"] == "success":
+                return SUCCESS
+            elif job["result"] == "busted" or job["result"] == "testfailed":
```
Should "testfailed" be WARNING? (Orange job on TH)

vaibhavmagarwal replied:
No, It will be FAILED. Because even if a test fails intermittently (ORANGE Job), that job is considered a failure. When a WARNING appears, the job remains GREEN and thus have 'success' code.
Eg log: http://ftp.mozilla.org/pub/mozilla.org/mobile/try-builds/bcampen@mozilla.com-28ab1ce9608e/try-android-api-9/try_ubuntu64_vm_mobile_test-mochitest-14-bm54-tests1-linux64-build363.txt.gz
Search for 'Warnings' in it, there are many, but the job remains green https://treeherder.mozilla.org/#/jobs?repo=try&revision=28ab1ce9608e

I have asked :camd, and he said that treeherder had no way to determine if there is a warning due to this.

armenzg commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+                return SUCCESS
+            elif job["result"] == "busted" or job["result"] == "testfailed":
+                return FAILURE
+            elif job["result"] == "skipped":
```
What does "skipped" represent?
Could you add a comment?

vaibhavmagarwal replied:
I will add a link to the possible statuses in treeherder: https://github.com/mozilla/treeherder/blob/master/treeherder/etl/buildbot.py#L7

armenzg commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,216 @@
+
+LOG = logging.getLogger('mozci')
+PENDING, RUNNING, COALESCED, UNKNOWN = range(-4, 0)
+SUCCESS, WARNING, FAILURE, SKIPPED, EXCEPTION, RETRY, CANCELLED = range(7)
```
I don't know how I feel about these constants being redefined in here when they already are defined in buildapi.py. I will think a bit more about it as I review the rest of the file.
We probably want to remove it from buildapi.py.
I would like to keep the comment from buildapi.py so we don't loose it.

armenzg commented on mozci/sources/buildapi.py:
```diff
@@ -196,64 +194,8 @@ def valid_revision(repo_name, revision):
-        return JOBS_CACHE[(repo_name, revision)]
-
+def query_for_jobs(repo_name, revision):
+    """ Query Buildapi for jobs """
```
Can you please keep the original comment from query_jobs_schedule?

armenzg replied:
Can you also keep the original name of the function? It helps those who are familiar with buildapi's terminology (since there is scheduling information from the Scheduler DB and then there is status information from the status DB).

vaibhavmagarwal replied:
This function is also defined in TreeherderApi class. This function returns all the jobs for a particular revision. If I change it here, I also need to change it in TreeherderApi class. Does changing it to query_jobs_schedule make sense in that case?

armenzg replied:
I can't see it defined for TreeherderApi class.
I see get_all_jobs() but not query_for_jobs().

vaibhavmagarwal replied:
Hmm, right. I myself got confused with so many names :)
I'll change the function name.

armenzg commented on mozci/sources/buildapi.py:
```diff
@@ -262,8 +204,7 @@ def query_jobs_schedule(repo_name, revision):
 
-    JOBS_CACHE[(repo_name, revision)] = req.json()
-    return JOBS_CACHE[(repo_name, revision)]
+    return req.json()
```
I believe we are loosing in here the ability to cache jobs.

armenzg replied:
At this moment, this looks to me like a regression.

vaibhavmagarwal replied:
We aren't. We are checking for the cache in get_all_jobs() here: https://github.com/armenzg/mozilla_ci_tools/pull/257/files#diff-c98ed40e085d771e97c54278bf4a8d64R49 , I have just added the core client part to buildapi and kept the caching part in query_jobs. Does it make sense?

armenzg replied:
It makes sense now.

armenzg commented:
@vaibhavmagarwal my apologies about this but this has been so far the most difficult code review in my career. It is a rather complex change and very prone to break so my apologies for this ardious process.

One major change (and hopefully the last major one) is that I would like as much of the original code of buildapi.py back into it. What I would like is the method sof the class BuildApi (the ones implementing the parent class) to stay but call explicitely code in buildapi.py.

This comes from the regression I see in query_jobs_schedule wrt to caching. It will also make the review process much easier since the buildapi module would stay as close as possible to its original status + we won't loose history in this rewrite.

vaibhavmagarwal commented:
Armen, I think I found your suggestion from last week about moving the "core" functionality of buildapi-client inside buildapi.py correct. Think about the Treeherder-client. It is just a thin wrapper around treeherder api, and does not deal with cache and stuff. I would like to have a similar functionality inside buildapi.py, that is, just using requests library get the response from buildapi. And we could keep the cache and other things that do not call self-serve url directly inside BuildApi class inside query_jobs.py. This will give a similar situation to TreeherderApi class, with the core calling functionality inside Treeherder-client for a general use case, and we have functions for it in query_jobs.py for our specific use case. Let me know if I am making sense, I know that this is a difficult decision, else I would change what you have said of moving the cache inside buildapi.

armenzg commented:
@vaibhavmagarwal thank you for explaining this to me.
It was hard to see at first. I need Alice to double check on this since it is a big change.
Please have a look at the other minor comments I made (for instance duplicating the constants).
Please wait to make changes until Alice has a chance to reply so we all agree.

@adusca vaibhav is proposing to make buildapi.py a module that *only* does basic actions on releng's self-serve/buildapi (similar to treeherder's python client).
On the other hand, he wants to move into query_jobs.py any handling of job caches.
I suggest only looking at the buildapi.py and query_jobs.py changes.

adusca commented:
@vaibhavmagarwal's proposal makes sense to me.

Since our unit test coverage is so low and we don't have performance tests for mozci I think we should test this with different modes to make sure everything is OK (for example, --coalesced mode takes between 30 and 40 minutes without caching but less then 2 minutes with caching, so we can see if anything went wrong with cache by calling trigger.py --coalesced and seeing how long it takes).

vaibhavmagarwal commented:
@armenzg I have addressed the comments in your review. @adusca I tested the coalesced part locally, it is getting executed within 2 mins, so I think caching is working fine. 
Armen, I know you will be away for 2 days, will it be possible for you to merge this now, since this patch is almost (if not fully) ready and major decisions have been made, some small issues/comments can be addressed afterwards too. I say this because of possible bitrot for this patch since Friday is a US holiday.

chmanchester commented:
@vaibhavmagarwal asked for some feedback from me on this. I'm going to take a look, but a lot of my feedback probably doesn't need to block merging this given the amount of review the PR has already seen.

chmanchester commented on mozci/mozci.py:
```diff
@@ -16,24 +16,35 @@
-    return matching_jobs
+    def wrapper(*args, **kwargs):
+        global QUERY_SOURCE
+        if QUERY_SOURCE is None:
```
Just for reference, the "global" statement is only necessary when modifying the value of a global variable, not when reading it.

chmanchester commented on mozci/mozci.py:
```diff
@@ -232,8 +245,9 @@ def query_jobs_buildername(buildername, revision):
-    all_jobs = buildapi.query_jobs_schedule(repo_name, revision)
-    jobs = _matching_jobs(buildername, all_jobs)
+    query_api = BuildApi()
+    all_jobs = query_api.get_all_jobs(repo_name, revision)
```
Is the data this function is getting only available from buildapi? I don't think I understand the purpose of this function.

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+        pass
+
+    @abstractmethod
+    def get_buildapi_request_id(self, repo_name, job):
```
Can we just refer to it as request_id so things seem more sane when we aren't getting it from buildapi?

vaibhavmagarwal replied:
There is a request_id used internally in treeherder too which we don't want, so I wanted to make it clear that we wanted the request_id of buildapi. Buildapi request_id can be retrieved from self-serve as well as treeherder. Therefore I named this get_buildapi_request_id(). Does get_request_id make more sense?

chmanchester replied:
In that case we're just victims of the existing conflict, we can keep it as get_buildapi_request_id.

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+            JOBS_CACHE[(repo_name, revision)] = \
+                buildapi.query_jobs_schedule(repo_name, revision)
+
+        return JOBS_CACHE[(repo_name, revision)]
```
I think this is faithful to the prior behavior, but when is this cache invalidated? It looks as though if I were to call this method once for a repo and revision, I would get the same response every subsequent time. This would be a problem if I were trying to use this function for triggerbot (some jobs finish before others are scheduled, so the count of builders would be persistently low for try pushes hitting lots of platforms).

armenzg replied:
We can follow up on this. Could you please file an issue?

The way that mozci has been used these days is by calling a script which takes a snapshot of the data at the moment and make decisions based on it.
mozci could not have any new information for at least a minute (when it shows up on builds-4hr.js).

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+
+    def get_job_status(self, job):
+        """Helper to determine the scheduling status of a job from self-serve."""
+        if not ("status" in job):
```
Might be cleaner as ```if "status" not in job:```

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+        status_data = query_job_data(req["complete_at"], req["request_id"])
+        return status_data["properties"]["revision"][0:12] != req["revision"][0:12]
+
+    def find_all_jobs_by_status(self, repo_name, revision, status):
```
This API is a little funny to me: find_all_jobs_by_status returns a list of matching jobs by repo and rev, but other methods like get_matching_jobs require passing the result of get_all_jobs. It might be nice to make this consistent (it also might hep clarify the caching situation for callers).

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+    __metaclass__ = ABCMeta
+
+    def __init__(self, query_source):
+        self.query_source = query_source
```
This member variable isn't read anywhere.

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+            revision_id = results[0]["id"]
+            all_jobs = self.treeherder_client.get_jobs(repo_name, count=2000,
+                                                       result_set_id=revision_id)
+        return all_jobs
```
Looking at the treeherder client, I don't think the result here will have the same structure as the result if this method were called on an instance of BuildApi. Ideally a caller would be able to use the result of get_all_jobs uniformly regardless of the data source.

chmanchester commented on mozci/query_jobs.py:
```diff
@@ -0,0 +1,221 @@
+            elif job["result"] == "retry":
+                return RETRY
+            elif job["result"] == "usercancel":
+                return CANCELLED
```
I might convert this long if/elif passage to a dictionary:
```python
{
    "success": SUCCESS,
    "busted":  FAILURE,
    "testfailed": FAILURE,
    ...
}
```

and get the results from there.

vaibhavmagarwal replied:
neat idea! I'll do that

armenzg commented:
Hi @vaibhavmagarwal , my apologies. I should have mentioned that I intended to have a look at this in my morning even if my day was off.
Please merge once you address any remaining issues raised.
I think we addressed all major issues already.

Thanks Chris for the extra good feedback!

vaibhavmagarwal commented on mozci/mozci.py:
```diff
@@ -16,24 +16,26 @@
 
+# Default value of QUERY_SOURCE
+QUERY_SOURCE = BuildApi
 
```
After discussing with @chmanchester, I have made default value of QUERY_SOURCE to BuildApi instead of None. That removes the need for a decorator for functions. Even if we moved set_query_source function to query_jobs.py, we will still require a global object to be initialized here, and also will have to define duplicate functions. So having one function made more sense in this case.

vaibhavmagarwal commented:
I have addressed all the major changes in this patch, and am merging this. I will file issues for remaining comments raised by Chris.
